server/src/test/java/com/google/fhir/gateway/BearerAuthorizationInterceptorTest.java

```diff
@@ -391,4 +395,24 @@ public String postProcess(HttpResponse response) throws IOException {
         requestDetails.getParameters().get("param3"),
         arrayContainingInAnyOrder("param3-value2", "param3-value1"));
   }
+
+  @Test
+  public void shouldSendGzippedResponseWhenRequested() throws IOException {
+    testInstance = createTestInstance(true, null);
+    String responseJson = "{\"resourceType\": \"Bundle\"}";
+    JWTCreator.Builder jwtBuilder = JWT.create().withIssuer(TOKEN_ISSUER);
+    when(requestMock.getHeader("Authorization")).thenReturn("Bearer " + signJwt(jwtBuilder));
+    when(requestMock.getHeader("Accept-Encoding".toLowerCase())).thenReturn("gzip");
+    when(requestMock.getServer()).thenReturn(serverMock);
+    ServletRestfulResponse proxyResponseMock = new ServletRestfulResponse(requestMock);
+    when(requestMock.getResponse()).thenReturn(proxyResponseMock);
+    HttpServletResponse proxyServletResponseMock = new MockHttpServletResponse();
+    when(requestMock.getServletResponse()).thenReturn(proxyServletResponseMock);
```

Comment:
I am confused in the 2 types of response mocks being set.
One is a ServletRestfulResponse and and other is HttpServletResponse. Why are we setting 2 different responses for an authorize request.

Reply:
Yes, I can understand.
`requestMock.getResponse()` is an abstraction HAPI provides to access the response object which is of type `ServletRestfulResponse`. [Internally HAPI uses the `HttpServletResponse` object](https://github.com/hapifhir/hapi-fhir/blob/master/hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/servlet/ServletRestfulResponse.java#L77) to perform any response related operations for this wrapper class `ServletRestfulResponse`. We have to perform mocking at two levels: one with `requestMock.getResponse()` because this is [how we access the wrapper response object](https://github.com/google/fhir-gateway/blob/main/server/src/main/java/com/google/fhir/gateway/BearerAuthorizationInterceptor.java#L301) and write to it.  We also need to perform a deeper level mock for the [internal HAPI operations](https://github.com/hapifhir/hapi-fhir/blob/master/hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/servlet/ServletRestfulResponse.java#L77) to be performed successfully.

This complication arises from us mocking the request object. Had the object been not mock, and set by a server we would not have needed to do this levels of mocks.

Reply:
Thanks for the explanation. 
I think it will be good to add some of this explanation in the comments(maybe a link to HAPI implementation). It is very hard for someone to understand this in future.